File: common/src/main/java/generations/gg/generations/core/generationscore/world/dialogue/nodes/spawning/BlockEntityYawLogic.java
package generations.gg.generations.core.generationscore.world.dialogue.nodes.spawning;

import dev.architectury.utils.value.FloatSupplier;
import generations.gg.generations.core.generationscore.client.model.ModelContextProviders.AngleProvider;
import net.minecraft.core.BlockPos;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceKey;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.block.HorizontalDirectionalBlock;
import net.minecraft.world.level.block.entity.BlockEntityType;

import java.util.Optional;

public record BlockEntityYawLogic(ResourceKey<BlockEntityType<?>> blockEntity) implements YawLogic {
    public FloatSupplier createSupplier(Player player) {
        var world = player.level();
        var optional = BlockPos.withinManhattanStream(player.getOnPos(), 10, 10, 10).map(a -> world.getBlockEntity(a, BuiltInRegistries.BLOCK_ENTITY_TYPE.get(blockEntity)))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst();
        if (optional.isPresent()) {
            var blockEntity = optional.get();
            var angle = blockEntity instanceof AngleProvider angleProvider ? angleProvider.getAngle() : blockEntity.getBlockState().hasProperty(HorizontalDirectionalBlock.FACING) ? blockEntity.getBlockState().getValue(HorizontalDirectionalBlock.FACING).toYRot() : 0f;
            return () -> angle;
        }
        return () -> 0f;
    }

    public static BlockEntityYawLogic of(ResourceKey<BlockEntityType<?>> key) {
        return new BlockEntityYawLogic(key);
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/world/dialogue/nodes/spawning/BlockEntityYawLogic.kt
package generations.gg.generations.core.generationscore.world.dialogue.nodes.spawning

import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.architectury.utils.value.FloatSupplier
import generations.gg.generations.core.generationscore.client.model.ModelContextProviders.AngleProvider
import net.minecraft.core.BlockPos
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.resources.ResourceKey
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.block.HorizontalDirectionalBlock
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import java.util.*
import java.util.function.Predicate

@JvmRecord
data class BlockEntityYawLogic(val blockEntity: ResourceKey<BlockEntityType<*>>) : YawLogic {
    override fun createSupplier(player: Player): FloatSupplier {
        val world = player.getLevel()
        val optional =
            BlockPos.withinManhattanStream(player.onPos, 10, 10, 10).map { a -> world.getBlockEntity(a, blockEntity.value())
            }
                .filter { obj -> obj.isPresent }
                .map { obj -> obj.get() }
                .findFirst()
        if (optional.isPresent) {
            val blockEntity = optional.get()
            val angle = if (blockEntity is AngleProvider) blockEntity.angle else if (blockEntity.blockState.hasProperty(
                    HorizontalDirectionalBlock.FACING
                )
            ) blockEntity.blockState.getValue(HorizontalDirectionalBlock.FACING).toYRot() else 0f
            return FloatSupplier { angle }
        }
        return FloatSupplier { 0f }
    }

    companion object {
        fun of(key: ResourceKey<BlockEntityType<*>>): BlockEntityYawLogic {
            return BlockEntityYawLogic(key)
        }
    }
}